### ISSUE TITLE
SshProxy<T> sudo init not working on XenServer/XCP-ng

### ISSUE THREAD

**Issue by jefflill** [nforgeio/neonKUBE#926]
I recently refactored how disabling sudo password prompts work for `SshProxy<T>`.  I tested this VMs provisioned via our Ubuntu 20.04.x templates on both Hyper-V and XenServer/XCP-ng (an exisiting instance).

XenServer/XCP-ng provisioning operations are now failing on a brand-new XCP-ng installation:
```
*** STATUS: connecting...
*** WAITFORBOOT: CONNECT TO [10.0.0.5:22]
*** STATUS: prepare: user folders
START: sudo bash -c 'if [ -f "/etc/sshproxy-init" ] ; then exit 0; else exit 1; fi'
STDERR
    sudo: sorry, you must have a tty to run sudo
END [ERROR=1]
*** STATUS: prepare: sudo
*** STATUS: connected
*** Downloading: /etc/os-release
START: sudo bash -c 'cp /etc/os-release /home/root/.neonkube/download/os-release-e31e0dcc-7b28-4267-9f61-a33793831cd0'
STDERR
    sudo: sorry, you must have a tty to run sudo
END [ERROR=1]
*** ERROR: *** ERROR Downloading: [SshProxyException]: [exitcode=1]: sudo: sorry, you must have a tty to run sudo

*** STACK:    at Neon.Kube.SshProxy`1.Download(String source, Stream output) in C:\src\neonKUBE\Lib\Neon.Kube\Proxies\SshProxy.cs:line 1538
START: sudo bash -c 'if [ -f "/home/root/.neonkube/download/os-release-e31e0dcc-7b28-4267-9f61-a33793831cd0" ] ; then rm "/home/root/.neonkube/download/os-release-e31e0dcc-7b28-4267-9f61-a33793831cd0" ; fi'
STDERR
    sudo: sorry, you must have a tty to run sudo
END [ERROR=1]
*** STATUS: [x] DONE
*** STATUS: check: virtual machines
START: xe vm-list params=all
END [OK]
*** STATUS: [x] DONE
*** STATUS: check: template
START: xe template-list params=all
END [OK]
*** STATUS: download: vm template (slow)
START: xe sr-list params=all
END [OK]
START: xe vm-import url=http://s3-us-west-2.amazonaws.com/neonforge/kube/xenserver-ubuntu-20.04.latest.xva sr-uuid=80273954-87d9-5ac4-8610-11953ce1f5b8
END [OK]
START: xe template-list params=all
END [OK]
START: xe template-param-set uuid=8aa4cc12-b9cf-5d9a-4285-ebdfb5488086 name-label=neon-ubuntu-20.04.latest
END [OK]
START: xe template-list params=all
END [OK]
*** STATUS: [x] DONE
START: xe host-list
END [OK]
START: xe host-param-list uuid=84f17839-6c52-4a90-8131-be3eadaf9778
END [OK]
*** STATUS: [jeff-master-0]: create: virtual machine
START: xe template-list params=all
END [OK]
START: xe sr-list params=all
END [OK]
START: xe vdi-list
END [OK]
START: xe vm-install template=neon-ubuntu-20.04.latest new-name-label=jeff-master-0 sr-uuid=80273954-87d9-5ac4-8610-11953ce1f5b8
END [OK]
START: xe vm-param-set uuid=3a8e1412-6285-1342-e940-06dd3454d84a VCPUs-at-startup=4 VCPUs-max=4
END [OK]
START: xe pool-list
END [OK]
START: xe pool-param-list uuid=f9a8ecae-bd48-d4a6-5656-955da2903e38
END [OK]
START: xe pool-param-set uuid=f9a8ecae-bd48-d4a6-5656-955da2903e38 other-config:auto_poweron=true
END [OK]
START: xe vm-param-set uuid=3a8e1412-6285-1342-e940-06dd3454d84a other-config:auto_poweron=true
END [OK]
START: xe vm-memory-limits-set uuid=3a8e1412-6285-1342-e940-06dd3454d84a dynamic-max=2147483648 dynamic-min=2147483648 static-max=2147483648 static-min=2147483648
END [OK]
START: xe vm-disk-list uuid=3a8e1412-6285-1342-e940-06dd3454d84a
END [OK]
START: xe vdi-resize uuid=5af9d730-2dee-43d8-83e1-61aec2456efd disk-size=10737418240
END [OK]
START: xe vm-list params=all
END [OK]
START: sudo bash -c 'mkdir /var/opt/neon-temp-sr'
STDERR
    sudo: sorry, you must have a tty to run sudo
END [ERROR=1]
START: sudo bash -c 'mkdir /var/opt/neon-temp-sr/68a1157b-b1af-43ad-a523-08c0cf08a045'
STDERR
    sudo: sorry, you must have a tty to run sudo
END [ERROR=1]
*** STATUS: ERROR[1]
*** ERROR: [RemoteCommandException]: [exitcode=1]: sudo bash -c 'export PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/snap/bin:/lib/neonkube/bin:/lib/neonkube/setup && mkdir /var/opt/neon-temp-sr/68a1157b-b1af-43ad-a523-08c0cf08a045'

*** ERROR: [RemoteCommandException]: [exitcode=1]: sudo bash -c 'export PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/snap/bin:/lib/neonkube/bin:/lib/neonkube/setup && mkdir /var/opt/neon-temp-sr/68a1157b-b1af-43ad-a523-08c0cf08a045'

*** STACK:    at Neon.Kube.SshProxy`1.RunCommand(String command, RunOptions runOptions, Object[] args) in C:\src\neonKUBE\Lib\Neon.Kube\Proxies\SshProxy.cs:line 2672
   at Neon.Kube.SshProxy`1.SudoCommand(String command, RunOptions runOptions, Object[] args) in C:\src\neonKUBE\Lib\Neon.Kube\Proxies\SshProxy.cs:line 2828
   at Neon.XenServer.XenClient.CreateTempIso(String isoPath, String srName) in C:\src\neonKUBE\Lib\Neon.XenServer\XenClient.cs:line 365
   at Neon.Kube.XenServerHostingManager.ProvisionVirtualMachines(SshProxy`1 xenSshProxy) in C:\src\neonKUBE\Lib\Neon.Kube.XenServer\XenServerHostingManager.cs:line 401
   at Neon.Kube.XenServerHostingManager.<Provision>b__14_2(SshProxy`1 node, TimeSpan stepDelay) in C:\src\neonKUBE\Lib\Neon.Kube.XenServer\XenServerHostingManager.cs:line 234
   at Neon.Kube.SetupController`1.<>c__DisplayClass39_0.<PerformStep>b__2(SshProxy`1 node) in C:\src\neonKUBE\Lib\Neon.Kube\Setup\SetupController.cs:line 532
```

Note these errors: **sudo: sorry, you must have a tty to run sudo**

Note that XenServer/XCP-ng are running on **Centos**, not **Ubuntu**.  The existing server had already been configured to disable sudo password prompts (from previous connections) whereas the brand new server needs to be configured.  I must need to do something different for Centos.

**Comment by jefflill** [comment]
Here's a link discussing this:

https://www.shell-tips.com/linux/sudo-sorry-you-must-have-a-tty-to-run-sudo/

It looks like Centos/RHEL shipped with a `/etc/sudoers` file that requires a TTY which people complained about back in 2014 as non-standard and not really helpful:

https://bugzilla.redhat.com/show_bug.cgi?id=1020147#c9

This was fixed back in 2014, but who knows whether XenServer/XCP-ng ever picked up the new version.

**Comment by jefflill** [comment]
I recently removed the `SshProxt<T>` code that used SSH.NET `ShellStream` to tweak these values:
```
// We need to make sure requiretty is turned off and that sudo can execute commands without a password.
// The sleeps are required when using the ShellStream, otherwise the commands won't work.

using (var sh = sshClient.CreateShellStream("terminal", 80, 40, 80, 40, 1024))
{
    sh.WriteLine("grep -qxF 'Defaults !requiretty' /etc/sudoers.d/notty || echo 'Defaults !requiretty' >> /etc/sudoers.d/notty");
    Thread.Sleep(500);
    sh.WriteLine("grep -qxF '%sudo    ALL=NOPASSWD: ALL' /etc/sudoers.d/nopasswd || echo '%sudo    ALL=NOPASSWD: ALL' >> /etc/sudoers.d/nopasswd");
    Thread.Sleep(500);
}
```
...so we've been missing with security all along.

This code is no longer required for Ubuntu servers and it wouldn't quite work for XenServer/XCP-ng because CentOS doesn't ship with the `/etc/sudoers.d/notty` and `/etc/sudoers.d/nopasswd` files.  Rather than hardcoding a solution that works just for this scenario, I'm going to add an optional action parameter to `SshProxy<T>.Connect()` and `WaitForBoot()` that will be called immediately after the SSH connection is established such that the caller can take the appropriate actions.

**Comment by jefflill** [comment]
I ended up just handling this in `SshProxt<T>` code rather than exposing a custom action to keep things simple.

**Comment by jefflill** [comment]
**FIXED** for v2.2.0
